src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb

```diff
@@ -24,26 +24,79 @@ def self.compile(compile_options)
       end
 
       def self.compile_ast(code, compile_options)
+        # TODO: move requires to optimal place and guard against FOSS not having bolt lib code
+        # CODEREVIEW: where should these requires be, and what should happen if they are not found?
+        require 'bolt'
+        require 'bolt/target'
+        require 'bolt/apply_inventory'
+        require 'bolt/apply_target'
+        require 'bolt/apply_result'
+        require 'addressable'
+
+        Puppet[:rich_data] = true
         Puppet[:node_name_value] = compile_options['certname']
 
+        # TODO: again, error handling/warning here when this is not supplied
+        boltlib_path = Array(compile_options.dig('options', 'boltlib') || '')
         # Use the existing environment with the requested name
         Puppet::Pal.in_environment(compile_options['environment'],
+                                   pre_modulepath: boltlib_path,
                                    envpath: Puppet[:environmentpath],
-                                   facts: compile_options['facts']['values'],
-                                   variables: compile_options['variables']['values']) do |pal|
-          Puppet.lookup(:pal_current_node).trusted_data = compile_options['trusted_facts']['values']
-
-          # This compiler has been configured with a node containing
-          # the requested environment, facts, and variables, and is used
-          # to compile a catalog in that context from the supplied AST.
-          pal.with_catalog_compiler do |compiler|
-            # We have to parse the AST inside the compiler block, because it
-            # initializes the necessary type loaders for us.
-            ast = Puppet::Pops::Serialization::FromDataConverter.convert(code)
-
-            compiler.evaluate(ast)
-            compiler.compile_additions
-            compiler.catalog_data_hash
+                                   facts: compile_options['facts']['values']) do |pal|
+          # No conn info for PE targets can just use empty hash if 
+          # https://github.com/puppetlabs/bolt/pull/1770 is accepted
+          fake_config = {
+            'transport' => 'redacted',
+            'transports' => {
+              'redacted' => 'redacted'
+            }
+          }
+          bolt_inv = Bolt::ApplyInventory.new(fake_config)
+          Puppet.override(bolt_inventory: bolt_inv) do
```

Comment:
What would happen if we didn't override this? iiuc, this will be `nil` in Puppet Server, does `nil` not for a "fake_config".

Reply:
We need the bolt inventory to support methods on Target objects. A bolt Target is a wrapper around a bolt inventory. Without an inventory backing a target, things like $target.facts will not work.